[GLib] Expose more GStreamer paths in sandbox

philn opened the pull request:
#### 47a42208eb062b7170cbcfc624a7752f44fc836e
<pre>
[GLib] Expose more GStreamer paths in sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=258403">https://bugs.webkit.org/show_bug.cgi?id=258403</a>

Reviewed by Michael Catanzaro.

Ensure the paths defined by GST_REGISTRY, GST_DEBUG_FILE and GST_DEBUG_DUMP_DOT_DIR are writable in
the WebProcess sandbox. A typo was also fixed in the GST_INSTALL_PLUGINS_HELPER environment variable
path and also now expose the ptp helper in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::environmentVariableValue):
(WebKit::bindGStreamerData):

Canonical link: <a href="https://commits.webkit.org/265773@main">https://commits.webkit.org/265773@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b704108f699f11c00874e1cbe05b0412c954187e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12071 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12444 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13517 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11309 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14460 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12052 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/14168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12860 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/10050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13939 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10152 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11230 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10937 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11347 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9382 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10522 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2854 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11203 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -392,20 +396,42 @@ static void bindGStreamerData(Vector<CString>& args)
             GUniquePtr<char> gstData(g_build_filename(g_get_user_data_dir(), "gstreamer-1.0", nullptr));
             bindIfExists(args, gstData.get());
         }
     }
```
It looks like user plugins are only bound if the two environment variables are not defined?

Can you decide what is the correct intended behavior and just fix it please? Seems weird to fix everything else except for this.

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -392,20 +396,42 @@ static void bindGStreamerData(Vector<CString>& args)
+    // The plugin scanner needs write permissions in the parent directory of GST_REGISTRY in order to
+    // write the registry file.
+    GUniquePtr<char> defaultRegistryPath(g_build_filename(g_get_user_cache_dir(), "gstreamer-1.0", nullptr));
+    const char* registryPath = g_getenv("GST_REGISTRY") ?: defaultRegistryPath.get();
```
Please don't use `?:`. I didn't even know about this weird language extension until I saw it here. It's not used outside this function, so you can get rid of the two preexisting uses in this patch.

philn replied:
I also discovered it in this file, it's called "Elvis operator": https://en.wikipedia.org/wiki/Elvis_operator

What's wrong with it? Why should it be avoided?

mcatanzaro replied:
It's a GCC language extension, not standard C++.

I'm surprised it works, because we use `--std=c++20` rather than `--std=gnu++20`.

philn replied:
Ah well... So long, Elvis.

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -392,20 +396,42 @@ static void bindGStreamerData(Vector<CString>& args)
-    bindIfExists(args, gstCache.get(), BindFlags::ReadWrite);
+    // The plugin scanner needs write permissions in the parent directory of GST_REGISTRY in order to
+    // write the registry file.
+    GUniquePtr<char> defaultRegistryPath(g_build_filename(g_get_user_cache_dir(), "gstreamer-1.0", nullptr));
```
Hmmm, well if write permissions are truly required, then it's required, but: if you give write permission here, then host GStreamer MUST be robust to malicious files in its cache. Hopefully this is actually true? Otherwise, it's a sandbox escape. The whole sandbox is only as good as the weakest link.

Maybe it would make sense to use a separate registry cache somewhere where the host will not accidentally use it, and have bwrap set the GST_REGISTRY environment variable to point to it? Then the write permissions are much less risky, right?

(In contrast, the write permissions for debug environment variables are not so scary because you know that by setting the environment variable, you definitely want GStreamer to write to that location.)

philn replied:
You mean, if GST_REGISTRY is set, we would re-set it to a different "safe" hardcoded path?

mcatanzaro replied:
Hm, no. If it's set, then you really want to use a particular registry path, and we should respect that. That's an explicit request by the user to allow the web process to use the registry.

I was thinking "if it's unset, then set it to someplace safe where the host will not see it," but thinking this through more, we could accomplish the same by just not mounting the default registry path, right? And that's already the current behavior. Do you agree? We should mount the registry path only if the environment variable is set?

philn replied:
OK, I think this is addressed in the latest version. Thanks for the review! I'll land this in a couple days, giving time to @TingPing to also give feedback.

TingPing replied:
As long as these are intended for developers it sounds fine to me.

mcatanzaro reviewed (approved):
Instead of the title `[WPE] WebProcess sandbox is too restrictive` how about something like `[GLib] Expose more GStreamer paths in sandbox`?

TingPing commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -172,13 +172,16 @@ enum class BindFlags {
+        return true;
     }
+
+    return false;
```
I expect this change will cause breakage.

Previously what it did was:

1. Find where a symlink pointed, and mount where it points
2. Continued on to mount the symlink path itself (in `bindIfExists()`)

Now it will be broken if the symlink is in a path not already mounted as step 2 never happens.

philn replied:
I'll revert this then :)

philn committed:
[GLib] Expose more GStreamer paths in sandbox
https://bugs.webkit.org/show_bug.cgi?id=258403

Reviewed by Michael Catanzaro.

Ensure the paths defined by GST_REGISTRY, GST_DEBUG_FILE and GST_DEBUG_DUMP_DOT_DIR are writable in
the WebProcess sandbox. A typo was also fixed in the GST_INSTALL_PLUGINS_HELPER environment variable
path and also now expose the ptp helper in the sandbox.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::environmentVariableValue):
(WebKit::bindGStreamerData):

Canonical link: https://commits.webkit.org/265773@main
